File: rust/54.spiral-matrix.rs
/*
 * @lc app=leetcode id=54 lang=rust
 *
 * [54] Spiral Matrix
 */

// @lc code=start
impl Solution {
    pub fn spiral_order(matrix: Vec<Vec<i32>>) -> Vec<i32> {
        let (m, n) = (matrix.len(), matrix[0].len());
        let mut result = Vec::with_capacity(m * n);
        let (mut top, mut right, mut bottom, mut left) = (0, n, m, 0);
        loop {
            if left == right {
                break;
            }
            result.extend(&matrix[top][left..right]);
            top += 1;
            if top == bottom {
                break;
            }
            result.extend(matrix[top..bottom].iter().map(|row| row[right - 1]));
            right -= 1;
            if left == right {
                break;
            }
            result.extend(matrix[bottom - 1][left..right].iter().rev());
            bottom -= 1;
            if top == bottom {
                break;
            }
            result.extend(matrix[top..bottom].iter().rev().map(|row| row[left]));
            left += 1;
        }
        result
    }
}
// @lc code=end

File: javascript/54.spiral-matrix.js
/*
 * @lc app=leetcode id=54 lang=javascript
 *
 * [54] Spiral Matrix
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    let top = 0, right = matrix[0].length, bottom = matrix.length, left = 0;
    const result = [];
    while (1) {
      if (left === right) break;
      for (let j = left; j < right; j++) {
        result.push(matrix[top][j]);
      }
      top++;
      if (top === bottom) break;
      for (let i = top; i < bottom; i++) {
        result.push(matrix[i][right - 1]);
      }
      right--;
      if (left === right) break;
      for (let j = right - 1; j >= left; j--) {
        result.push(matrix[bottom - 1][j]);
      }
      bottom--;
      if (top === bottom) break;
      for (let i = bottom - 1; i >= top; i--) {
        result.push(matrix[i][left]);
      }
      left++;
    }
    return result;
};
// @lc code=end

